### PR TITLE
[Admin]Admin choose channel in product show page

### PR DESCRIPTION
**Pull request by Tomanhez**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | yes
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

**Before clicking:**
Screenshot 2019-07-01 at 10 29 19

**After clicking in several channels:**
Screenshot 2019-07-01 at 10 29 51

**One Channel**
Screenshot 2019-07-03 at 13 42 41

**All channels are disabled in product**
Screenshot 2019-07-04 at 10 01 31

### CODE REVIEW

File: features/product/viewing_product_in_admin_panel/viewing_product_for_particular_channel.feature [hunk 1]
```diff
@@ -0,0 +1,20 @@
+@viewing_products
+Feature: Viewing a product for a particular channel
+    In order to choosing view a product in each channel
```
Comment (GSadee) [inline features/product/viewing_product_in_admin_panel/viewing_product_for_particular_channel.feature]:
```suggestion
    In order to quickly view a product in a chosen channel
```

File: features/product/viewing_product_in_admin_panel/viewing_product_for_particular_channel.feature [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+        And the store classifies its products as "Shield" and "Equipment"
+        And the store has a product "Iron shield" priced at "$20.00" in "UnitedStates" channel
+        And this product is available in "Europe" channel
+        And I am logged in as an administrator
+        And I am browsing products
+
+    @ui
+    Scenario: Viewing product shop page for particular channel
+        When I access "Iron shield" product page
+        And I specify this product to viewing in "Europe" channel
```
Comment (GSadee) [inline features/product/viewing_product_in_admin_panel/viewing_product_for_particular_channel.feature]:
```suggestion
        And I view this product in "Europe" channel
```

File: src/Sylius/Behat/Context/Setup/ChannelContext.php [hunk 3]
```diff
@@ -103,12 +103,15 @@ public function storeOperatesOnASingleChannel($currencyCode = null)
      * @Given /^the store(?:| also) operates on (?:a|another) channel named "([^"]+)"$/
      * @Given /^the store(?:| also) operates on (?:a|another) channel named "([^"]+)" in "([^"]+)" currency$/
      * @Given the store operates on a channel identified by :code code
+     * @Given the store (also) operates on a channel named :channelName with hostname :hostname
      */
-    public function theStoreOperatesOnAChannelNamed($channelName, $currencyCode = null)
+    public function theStoreOperatesOnAChannelNamed($channelName, $currencyCode = null, string $hostname = null)
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Setup/ChannelContext.php]: If you are modifying this line, you could also add missing parameter types

File: src/Sylius/Behat/Context/Ui/Admin/ProductShowPageContext.php [hunk 4]
```diff
@@ -334,4 +349,14 @@ public function iShouldSeeAttributeWithValueInLocale(string $attribute, string $
     {
         Assert::true($this->attributesElement->hasAttributeInLocale($attribute, $locale, $value));
     }
+
+    /**
+     * @Then /^I should see (this product) in ("([^"]*)" channel)$/
+     */
+    public function iShouldSeeThisProductInChannel(ProductInterface $product, ChannelInterface $channel): void
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Admin/ProductShowPageContext.php]: I would be for moving this to a separate context and do not mix the admin and shop steps in one context

File: src/Sylius/Behat/Context/Ui/Admin/ProductShowPageContext.php [hunk 5]
```diff
@@ -334,4 +349,14 @@ public function iShouldSeeAttributeWithValueInLocale(string $attribute, string $
...
+    {
+        Assert::true(null !== strpos($this->shopProductShowPage->getCurrentUrl(), $channel->getHostname()));
+
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Admin/ProductShowPageContext.php]: Unnecessary blank line imo

File: src/Sylius/Behat/Page/Admin/Product/ShowPage.php [hunk 6]
```diff
@@ -32,10 +33,16 @@ public function getRouteName(): string
...
+    public function specifyChannel(ChannelInterface $channel): void
+    {
+        $this->getElement('scrolling_menu')->find('css', sprintf("a:contains('%s')", $channel->getName()))->click();
+    }
+
     protected function getDefinedElements(): array
     {
         return array_merge(parent::getDefinedElements(), [
             'product_name' => '#header h1 .content > span',
+            'scrolling_menu' => '.scrolling.menu',
```
Comment (GSadee) [inline src/Sylius/Behat/Page/Admin/Product/ShowPage.php]:
```suggestion
            'scrolling_menu' => '.scrolling.menu a:contains('%channelName%')',
```

File: src/Sylius/Behat/Page/Admin/Product/ShowPage.php [hunk 7]
```diff
@@ -32,10 +33,16 @@ public function getRouteName(): string
         return 'sylius_admin_product_show';
     }
 
+    public function specifyChannel(ChannelInterface $channel): void
```
Comment (GSadee) [inline src/Sylius/Behat/Page/Admin/Product/ShowPage.php]: This method name does not seem to be appropriate for behaviour

File: features/product/viewing_product_in_admin_panel/viewing_product_for_particular_channel.feature [hunk 8]
```diff
@@ -0,0 +1,20 @@
+@viewing_products
+Feature: Viewing a product for a particular channel
```
Comment (Zales0123) [inline features/product/viewing_product_in_admin_panel/viewing_product_for_particular_channel.feature]: `Viewing a product in a particular channel`

File: src/Sylius/Bundle/AdminBundle/Resources/views/Product/Show/_header.html.twig [hunk 9]
```diff
@@ -11,10 +11,24 @@
...
+                <i class="share alternate icon"></i>
+                <span class="text">
                 {{ 'sylius.ui.show_product_in_shop_page'|trans }}
-            </a>
+                </span>
+                <div class="menu">
+                    <div class="scrolling menu">
+                        {% for channel in product.channels %}
+                            {% set url = channel.hostname is not null ? 'http://' ~ channel.hostname ~ path('sylius_shop_product_show', {'slug': product.slug, '_locale': channel.defaultLocale.code}) : url('sylius_shop_product_show', {'slug': product.slug, '_locale': channel.defaultLocale.code}) %}
+                                <a href="{{ url|raw }}" class="item">
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Product/Show/_header.html.twig]: Wrong indentation

File: features/product/viewing_product_in_admin_panel/viewing_product_for_particular_channel.feature (hunk 1 above)
Comment (Zales0123) [inline features/product/viewing_product_in_admin_panel/viewing_product_for_particular_channel.feature]: `In order to easily check product details as seen by a Customer`? cc @CoderMaggie

File: features/product/viewing_product_in_admin_panel/viewing_product_for_particular_channel.feature [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+    In order to choosing view a product in each channel
+    As an Administrator
+    I want to be able to view product shop page for particular channel
```
Comment (Zales0123) [inline features/product/viewing_product_in_admin_panel/viewing_product_for_particular_channel.feature]: `I want to be able to view product details in the particular channel`

File: features/product/viewing_product_in_admin_panel/viewing_product_for_particular_channel.feature [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+
+    Background:
+        Given the store operates on a channel named "UnitedStates" with hostname "goodcars.com"
```
Comment (Zales0123) [inline features/product/viewing_product_in_admin_panel/viewing_product_for_particular_channel.feature]: `United States`

File: features/product/viewing_product_in_admin_panel/viewing_product_for_particular_channel.feature [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+        And the store also operates on a channel named "Europe" with hostname "goodcars.eu"
+        And the store classifies its products as "Shield" and "Equipment"
```
Comment (Zales0123) [inline features/product/viewing_product_in_admin_panel/viewing_product_for_particular_channel.feature]: Make it more car related, the same with product 💃

File: features/product/viewing_product_in_admin_panel/viewing_product_for_particular_channel.feature [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+        And the store has a product "Iron shield" priced at "$20.00" in "UnitedStates" channel
+        And this product is available in "Europe" channel
+        And I am logged in as an administrator
+        And I am browsing products
```
Comment (Zales0123) [inline features/product/viewing_product_in_admin_panel/viewing_product_for_particular_channel.feature]: Is it needed?

Comment (Zales0123) [reply features/product/viewing_product_in_admin_panel/viewing_product_for_particular_channel.feature]: If yes, should be in a scenario

File: features/product/viewing_product_in_admin_panel/viewing_product_for_particular_channel.feature (hunk 2 above)
Comment (Zales0123) [inline features/product/viewing_product_in_admin_panel/viewing_product_for_particular_channel.feature]: `And I show this product in a "Europe" channel`

File: src/Sylius/Behat/Context/Ui/Admin/ProductShowPageContext.php [hunk 14]
```diff
@@ -37,6 +39,9 @@ final class ProductShowPageContext implements Context
     /** @var ShowPageInterface */
     private $productShowPage;
 
+    /** @var ShopShowPageInterface */
+    private $shopProductShowPage;
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Admin/ProductShowPageContext.php]: It should not be here, it's a shop page 🚀

File: src/Sylius/Behat/Page/Admin/Product/ShowPage.php (hunk 6 above)
Comment (Zales0123) [inline src/Sylius/Behat/Page/Admin/Product/ShowPage.php]: `'show_product_dropdown'`?

File: src/Sylius/Behat/Page/Admin/Product/ShowPage.php [hunk 15]
```diff
@@ -32,10 +33,16 @@ public function getRouteName(): string
...
+    {
+        $this->getElement('scrolling_menu')->find('css', sprintf("a:contains('%s')", $channel->getName()))->click();
```
Comment (Zales0123) [inline src/Sylius/Behat/Page/Admin/Product/ShowPage.php]: `->find('css', sprintf("a:contains('%s')", $channel->getName()))->click()` can be replaced with `->clickLink($channel->getName())`

File: src/Sylius/Behat/Page/Admin/Product/ShowPageInterface.php [hunk 16]
```diff
@@ -14,10 +14,13 @@
...
 use FriendsOfBehat\PageObjectExtension\Page\SymfonyPageInterface;
+use Sylius\Component\Core\Model\ChannelInterface;
 
 interface ShowPageInterface extends SymfonyPageInterface
 {
     public function getName(): string;
 
     public function isSimpleProductPage(): bool;
+
+    public function specifyChannel(ChannelInterface $channel): void;
```
Comment (Zales0123) [inline src/Sylius/Behat/Page/Admin/Product/ShowPageInterface.php]: `showProductInChannel`

Comment (Zales0123) [reply src/Sylius/Behat/Page/Admin/Product/ShowPageInterface.php]: And do we need the whole channel there? Maybe just a channel name would be enough?

File: src/Sylius/Bundle/AdminBundle/Resources/views/Product/Show/_header.html.twig [hunk 17]
```diff
@@ -11,10 +11,24 @@
...
+            <div class="ui floating dropdown labeled icon button">
+                <i class="share alternate icon"></i>
+                <span class="text">
                 {{ 'sylius.ui.show_product_in_shop_page'|trans }}
-            </a>
+                </span>
+                <div class="menu">
+                    <div class="scrolling menu">
+                        {% for channel in product.channels %}
+                            {% set url = channel.hostname is not null ? 'http://' ~ channel.hostname ~ path('sylius_shop_product_show', {'slug': product.slug, '_locale': channel.defaultLocale.code}) : url('sylius_shop_product_show', {'slug': product.slug, '_locale': channel.defaultLocale.code}) %}
```
Comment (Zales0123) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Product/Show/_header.html.twig]: I suppose we should not hardcode `http://` there (a website can be accessible from https, for example)

Comment (Tomanhez) [reply src/Sylius/Bundle/AdminBundle/Resources/views/Product/Show/_header.html.twig]: Have you any idea how to do it? If i delete 'http://'  i get error, I don't have better resolve but i suspect that hardcoding is not good way.

File: src/Sylius/Bundle/AdminBundle/Resources/views/Product/Show/_header.html.twig [hunk 18]
```diff
@@ -9,12 +9,36 @@
...
+    {% set protocol = 'http://' %}
+    {% set counter = 0 %}
+    {% for server in configuration.request.server %}
+        {% set counter = counter + 1 %}
+        {% if  counter == 5%}
+            {% if 'HTTPS' in server %}
+                {% set protocol = 'https://' %}
+            {% endif %}
+        {% endif %}
+    {% endfor %}
```
Comment (Tomanhez) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Product/Show/_header.html.twig]: cc @Zales0123 what are you thinking about this solve, its looks like cancer but i think that its working ??

File: features/product/viewing_product_in_admin_panel/viewing_product_for_particular_channel.feature [hunk 19]
```diff
@@ -0,0 +1,20 @@
+@viewing_products
+Feature: Viewing a product in a particular channel
+    In order to quickly view a product in a chosen channel
+    As an Administrator
+    I want to be able to view product details in the particular channel
```
Comment (CoderMaggie) [inline features/product/viewing_product_in_admin_panel/viewing_product_for_particular_channel.feature]: To be improved ;)

File: features/product/viewing_product_in_admin_panel/viewing_product_for_particular_channel.feature [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+        And the store classifies its products as "Cars" and "Equipment"
+        And the store has a product "Bugatti" priced at "$20.00" in "United States" channel
+        And this product is available in "Europe" channel
+        And I am logged in as an administrator
+
+    @ui
+    Scenario: Viewing product shop page for particular channel
+        Given I am browsing products
+        When I access "Bugatti" product page
+        And I show this product in a "Europe" channel
```
Comment (CoderMaggie) [inline features/product/viewing_product_in_admin_panel/viewing_product_for_particular_channel.feature]:
```suggestion
        And I show this product in the "Europe" channel
```

File: features/product/viewing_product_in_admin_panel/viewing_product_for_particular_channel.feature [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+        Then I should see this product in "Europe" channel
```
Comment (CoderMaggie) [inline features/product/viewing_product_in_admin_panel/viewing_product_for_particular_channel.feature]:
```suggestion
        Then I should see this product in the "Europe" channel in shop
```

File: features/product/viewing_product_in_admin_panel/viewing_product_for_particular_channel.feature [hunk 22]
```diff
@@ -0,0 +1,20 @@
...
+    In order to quickly view a product in a chosen channel
+    As an Administrator
+    I want to be able to view product details in the particular channel
+
+    Background:
+        Given the store operates on a channel named "United States" with hostname "goodcars.com"
+        And the store also operates on a channel named "Europe" with hostname "goodcars.eu"
+        And the store classifies its products as "Cars" and "Equipment"
+        And the store has a product "Bugatti" priced at "$20.00" in "United States" channel
+        And this product is available in "Europe" channel
```
Comment (CoderMaggie) [inline features/product/viewing_product_in_admin_panel/viewing_product_for_particular_channel.feature]:
```suggestion
        And this product is available in the "Europe" channel
```

File: src/Sylius/Bundle/AdminBundle/Resources/views/Product/Show/_header.html.twig (hunk 18 above)
Comment (lchrusciel) [reply src/Sylius/Bundle/AdminBundle/Resources/views/Product/Show/_header.html.twig]: Why is it needed? Can we just skip protocol and leave the browser to assign one of them?

File: src/Sylius/Bundle/AdminBundle/Resources/views/Product/Show/_header.html.twig [hunk 23]
```diff
@@ -9,12 +9,48 @@
...
+        <div class="middle aligned column">
+            <div class="ui right floated buttons">
+                <div class="ui floating dropdown labeled icon button">
+                    <i class="share alternate icon"></i>
+                    <span class="text">
+                    {{ 'sylius.ui.show_product_in_shop_page'|trans }}
+                    </span>
+                    <div class="menu">
+                        <div class="scrolling menu">
+                            {% for channel in product.channels %}
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Product/Show/_header.html.twig]:
```suggestion
                            {% for channel in product.channels if channel.enabled == true %}
```
And remove below if

File: src/Sylius/Bundle/AdminBundle/Resources/views/Product/Show/_header.html.twig [hunk 24]
```diff
@@ -9,12 +9,48 @@
...
-        <div class="ui right floated buttons">
-            <a class="ui labeled icon button {% if not product.enabled %} disabled {% endif %}" href="{{ path('sylius_shop_product_show', {'slug': product.slug, '_locale': product.translation.locale}) }}">
-                <i class="angle right icon"></i>
-                {{ 'sylius.ui.show_product_in_shop_page'|trans }}
-            </a>
+    {% set channelCounter = 0 %}
+    {% for channel in product.channels %}
+        {% set channelCounter = channelCounter + 1 %}
+    {% endfor %}
+    {% if channelCounter > 1 %}
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Product/Show/_header.html.twig]:
```suggestion
    {% if product.channels|length > 1 %}
```
And above foreach is not needed

File: src/Sylius/Bundle/AdminBundle/Resources/views/Product/Show/_header.html.twig [hunk 25]
```diff
@@ -9,12 +9,42 @@
...
+            <div class="ui right floated buttons">
+                <div class="ui floating dropdown labeled icon button">
+                    <i class="share alternate icon"></i>
+                    <span class="text">
+                    {{ 'sylius.ui.show_product_in_shop_page'|trans }}
+                    </span>
+                    <div class="menu">
+                        <div class="scrolling menu">
+                            {% for channel in product.channels if channel.enabled == true %}
+                                    {% set url = channel.hostname is not null ? 'http://' ~ channel.hostname ~ path('sylius_shop_product_show', {'slug': product.slug, '_locale': channel.defaultLocale.code}) : url('sylius_shop_product_show', {'slug': product.slug, '_locale': channel.defaultLocale.code}) %}
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Product/Show/_header.html.twig]: Wrong indentation

File: src/Sylius/Bundle/AdminBundle/Resources/views/Product/Show/_header.html.twig [hunk 26]
```diff
@@ -9,12 +9,42 @@
...
+                        </div>
+                    </div>
+                </div>
+            </div>
         </div>
-    </div>
+    {% else %}
+        <div class="middle aligned column">
+            <div class="ui right floated buttons">
+                {% for channel in product.channels %}
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Product/Show/_header.html.twig]: Change in the same way as you have changed foreach above

File: src/Sylius/Bundle/AdminBundle/Resources/views/Product/Show/_header.html.twig [hunk 27]
```diff
@@ -9,12 +9,42 @@
         </h1>
         {% include "@SyliusAdmin/Product/Show/_breadcrumb.html.twig" %}
     </div>
-    <div class="middle aligned column">
-        <div class="ui right floated buttons">
-            <a class="ui labeled icon button {% if not product.enabled %} disabled {% endif %}" href="{{ path('sylius_shop_product_show', {'slug': product.slug, '_locale': product.translation.locale}) }}">
-                <i class="angle right icon"></i>
-                {{ 'sylius.ui.show_product_in_shop_page'|trans }}
-            </a>
+    {% if product.channels|length > 1 %}
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Product/Show/_header.html.twig]: You don't check if channels are enabled for this product, maybe you can try to use twig filter?
https://twig.symfony.com/doc/2.x/filters/filter.html

File: src/Sylius/Bundle/AdminBundle/Resources/views/Product/Show/_header.html.twig (hunk 26 above)
Comment (GSadee) [reply src/Sylius/Bundle/AdminBundle/Resources/views/Product/Show/_header.html.twig]: However, I'm not sure if using foreach when you have one or zero channels is a correct solution here. What will happen if we do not have any channel?

File: src/Sylius/Behat/Context/Setup/ChannelContext.php [hunk 28]
```diff
@@ -103,12 +103,15 @@ public function storeOperatesOnASingleChannel($currencyCode = null)
      * @Given /^the store(?:| also) operates on (?:a|another) channel named "([^"]+)"$/
      * @Given /^the store(?:| also) operates on (?:a|another) channel named "([^"]+)" in "([^"]+)" currency$/
      * @Given the store operates on a channel identified by :code code
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Setup/ChannelContext.php]: According to code, it is soooo much not true. Can you open an issue with it

File: src/Sylius/Behat/Context/Setup/ChannelContext.php [hunk 29]
```diff
@@ -103,12 +103,15 @@ public function storeOperatesOnASingleChannel($currencyCode = null)
...
+     * @Given the store (also) operates on a channel named :channelName with hostname :hostname
      */
-    public function theStoreOperatesOnAChannelNamed($channelName, $currencyCode = null)
+    public function theStoreOperatesOnAChannelNamed(string $channelName, string $currencyCode = null, string $hostname = null): void
     {
         $channelCode = StringInflector::nameToLowercaseCode($channelName);
         $defaultData = $this->defaultChannelFactory->create($channelCode, $channelName, $currencyCode);
 
+        ($hostname !== null) ? $defaultData['channel'] = $defaultData['channel']->setHostname($hostname) : null;
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Setup/ChannelContext.php]:
```suggestion
		$defaultData['channel']->setHostname($hostname);
```

**Comment by lchrusciel** [comment]
Thank you, Tomasz! :1st_place_medal:

**Pull request merged**